File: deltafi-core/src/main/java/org/deltafi/core/services/api/model/Event.java
/*
 *    DeltaFi - Data transformation and enrichment platform
 *
 *    Copyright 2021-2023 DeltaFi Contributors <deltafi@deltafi.org>
 *
 *    Licensed under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License.
 *    You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software
 *    distributed under the License is distributed on an "AS IS" BASIS,
 *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *    See the License for the specific language governing permissions and
 *    limitations under the License.
 */
package org.deltafi.core.services.api.model;

import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Getter;
import org.apache.commons.text.StringEscapeUtils;

import java.util.List;

@Builder(builderMethodName = "hiddenBuilder")
@Getter
@AllArgsConstructor
public class Event {

    public enum Severity {
        ERROR("error"),
        WARN("warn"),
        INFO("info"),
        SUCCESS("success");

        public final String label;

        Severity(String label) {
            this.label = label;
        }
    }

    public final String summary;
    @Builder.Default
    public final Severity severity = Severity.INFO;
    @Builder.Default
    public final String source = "core";
    @Builder.Default
    public final boolean notification = false;
    @Builder.Default
    public final String content = null;

    // Just to make javadocs happy
    public static class EventBuilder {

        private static final String COLUMN_SEP = "|";
        private static final String HEADER_SEP = "-";
        public static final String ESCAPED_NEWLINE = "\\n";

        public EventBuilder addList(String title, List<String> items) {
            if(items != null && !items.isEmpty()) {
                String blob = title + "\\n* " + String.join("\\n* ", items) + ESCAPED_NEWLINE;
                if (content$set) {
                    content(content$value + ESCAPED_NEWLINE + blob);
                } else {
                    content(blob);
                }
            }
            return this;
        }

        public EventBuilder addTable(String title, List<String> headers, List<List<String>> rows) {
            if (rows != null && !rows.isEmpty()) {
                StringBuilder table = new StringBuilder(title);
                table.append(ESCAPED_NEWLINE);
                table.append(COLUMN_SEP).append(String.join(COLUMN_SEP, headers)).append(COLUMN_SEP).append(ESCAPED_NEWLINE);
                table.append(COLUMN_SEP);
                headers.forEach(header -> table.append(HEADER_SEP.repeat(header.length())).append(COLUMN_SEP));
                table.append(ESCAPED_NEWLINE);
                rows.forEach(row -> addRow(table, row));

                if (content$set) {
                    content(content$value + ESCAPED_NEWLINE + table);
                } else {
                    content(table.toString());
                }
            }
            return this;
        }

        private void addRow(StringBuilder content, List<String> row) {
            content.append(COLUMN_SEP);
            row.forEach(value -> content.append(StringEscapeUtils.escapeJson(value)).append(COLUMN_SEP));
            content.append(ESCAPED_NEWLINE);
        }

        public EventBuilder addJsonLog(String title, String jsonLog) {
            String blob = title + "\\n```json\\n" + StringEscapeUtils.escapeJson(jsonLog) +"```";
            if (content$set) {
                content(content$value + ESCAPED_NEWLINE + blob);
            } else {
                content(blob);
            }
            return this;
        }
    }

    public static EventBuilder builder(String summary) {
        return hiddenBuilder().summary(summary);
    }

    public String asJson() {
        if (content == null) {
            return String.format("""
                    {
                        "summary": "%s",
                        "source": "%s",
                        "severity": "%s",
                        "notification": %s
                    }""", summary, source, severity.label, notification ? "true" : "false");
        } else {
            return String.format("""
                    {
                        "summary": "%s",
                        "content": "%s",
                        "source": "%s",
                        "severity": "%s",
                        "notification": %s
                    }""", summary, content, source, severity.label, notification ? "true" : "false");
        }
    }

}

File: deltafi-analytics/internal/schema/event.go
/*
 *    DeltaFi - Data transformation and enrichment platform
 *
 *    Copyright 2021-2025 DeltaFi Contributors <deltafi@deltafi.org>
 *
 *    Licensed under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License.
 *    You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software
 *    distributed under the License is distributed on an "AS IS" BASIS,
 *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *    See the License for the specific language governing permissions and
 *    limitations under the License.
 */
/*
 * ABOUTME: Defines the analytics event and annotation schemas for Parquet storage.
 * ABOUTME: Uses normalized design to support late-arriving annotations.
 */
package schema

import "time"

// Event represents a core analytics event without annotations.
// Annotations are stored separately to support late-arriving updates.
type Event struct {
	DID        string `parquet:"did"`
	DataSource string `parquet:"data_source"`

	EventType    string    `parquet:"event_type"` // INGRESS, EGRESS, ERROR, FILTER, CANCEL
	EventTime    time.Time `parquet:"event_time,timestamp(millisecond)"`
	IngestTime   time.Time `parquet:"ingest_time,timestamp(millisecond)"`
	CreationTime time.Time `parquet:"-"` // DeltaFile creation time for hour-based partitioning (not stored in parquet)

	Bytes     int64 `parquet:"bytes"`
	FileCount int32 `parquet:"file_count"`

	FlowName    string `parquet:"flow_name,optional"`
	ActionName  string `parquet:"action_name,optional"`
	Cause       string `parquet:"cause,optional"`
	IngressType string `parquet:"ingress_type,optional"` // DATA_SOURCE, CHILD, SURVEY
}

// Annotation represents a key-value annotation for a DID.
// Stored separately from events to support late-arriving updates.
type Annotation struct {
	DID          string    `parquet:"did"`
	Key          string    `parquet:"key"`
	Value        string    `parquet:"value"`
	UpdateTime   time.Time `parquet:"update_time,timestamp(millisecond)"`
	CreationTime time.Time `parquet:"-"` // DeltaFile creation time for hour-based partitioning (not stored in parquet)
}

// PreAggregatedEvent is the intermediate stage between raw events and final aggregation.
// Preserves DID for annotation joins, partitioned by creation_time, has event_time bucket computed.
// This stage allows annotations to be joined efficiently before final aggregation by event_time.
type PreAggregatedEvent struct {
	DID            string    `parquet:"did"`                               // Preserved for annotation joins
	EventTimeBucket time.Time `parquet:"event_time_bucket,timestamp(millisecond)"` // 5-min bucket by EVENT time
	DataSource     string    `parquet:"data_source"`
	EventType      string    `parquet:"event_type"`
	FlowName       string    `parquet:"flow_name,optional"`
	ActionName     string    `parquet:"action_name,optional"`
	Cause          string    `parquet:"cause,optional"`
	IngressType    string    `parquet:"ingress_type,optional"`

	// Annotations joined from annotation files
	Annotations map[string]string `parquet:"annotations,optional"`

	EventCount     int64 `parquet:"event_count"`
	TotalBytes     int64 `parquet:"total_bytes"`
	TotalFileCount int64 `parquet:"total_file_count"`
}

// AggregatedEvent represents final rolled-up metrics for querying.
// Groups by 5-minute bucket + dimensions + annotations, drops DIDs, sums metrics.
// Partitioned by EVENT_TIME date for efficient time-range queries.
// Annotations stored as Parquet MAP type for efficient columnar storage and querying.
type AggregatedEvent struct {
	Bucket      time.Time `parquet:"bucket,timestamp(millisecond)"` // 5-min bucket by EVENT time
	DataSource  string    `parquet:"data_source"`
	EventType   string    `parquet:"event_type"`
	FlowName    string    `parquet:"flow_name,optional"`
	ActionName  string    `parquet:"action_name,optional"`
	Cause       string    `parquet:"cause,optional"`
	IngressType string    `parquet:"ingress_type,optional"` // DATA_SOURCE, CHILD, SURVEY

	// Annotations as MAP type - queryable as annotations['key'] in DuckDB
	Annotations map[string]string `parquet:"annotations,optional"`

	EventCount     int64 `parquet:"event_count"`
	TotalBytes     int64 `parquet:"total_bytes"`
	TotalFileCount int64 `parquet:"total_file_count"`
}

// EventRequest is the JSON structure received via HTTP POST for events
type EventRequest struct {
	DID          string            `json:"did"`
	DataSource   string            `json:"dataSource"`
	EventType    string            `json:"eventType"`
	EventTime    int64             `json:"eventTime"`    // Unix millis - when the event occurred
	CreationTime int64             `json:"creationTime"` // Unix millis - when the DeltaFile was created (used for partitioning)
	Bytes        int64             `json:"bytes"`
	FileCount    int32             `json:"fileCount"`
	FlowName     string            `json:"flowName,omitempty"`
	ActionName   string            `json:"actionName,omitempty"`
	Cause        string            `json:"cause,omitempty"`
	IngressType  string            `json:"ingressType,omitempty"`
	Annotations  map[string]string `json:"annotations,omitempty"`
}

// AnnotationRequest is the JSON structure for annotation updates
type AnnotationRequest struct {
	DID          string            `json:"did"`
	Annotations  map[string]string `json:"annotations"`
	CreationTime int64             `json:"creationTime,omitempty"` // DeltaFile creation time (Unix millis) for hour-based partitioning
}
